qiskit/circuit/classical/types/types.py

```diff
@@ -27,127 +21,10 @@
     "Uint",
 ]
 
-import typing
-
-
-class _Singleton(type):
-    """Metaclass to make the child, which should take zero initialization arguments, a singleton
-    object."""
-
-    def _get_singleton_instance(cls):
-        return cls._INSTANCE
-
-    @classmethod
-    def __prepare__(mcs, name, bases):  # pylint: disable=unused-argument
-        return {"__new__": mcs._get_singleton_instance}
-
-    @staticmethod
-    def __new__(cls, name, bases, namespace):
-        out = super().__new__(cls, name, bases, namespace)
-        out._INSTANCE = object.__new__(out)  # pylint: disable=invalid-name
-        return out
```

Comment:
I'm assuming we don't use this anymore because of how we've implemented `Type` from Rust. But I imagined this was used to discriminate against some types. Knowing that the other types are "stateless" the only one that wasn't singleton was/is `Unit` as it contains a `u16`.

Reply:
I know there's a test somewhere that asserts that the `Bool` type _is_ a singleton. But, our code doesn't actually rely on this anywhere from what I could tell.

...but, all of the types that were singletons in Python I've actually made to be singletons in Rust as well. You'll notice I'm caching these instances at the top of `types.rs` 🙂.

```rust
static BOOL_TYPE: GILOnceCell<Py<PyBool>> = GILOnceCell::new();
static DURATION_TYPE: GILOnceCell<Py<PyDuration>> = GILOnceCell::new();
static FLOAT_TYPE: GILOnceCell<Py<PyFloat>> = GILOnceCell::new();
```

I thought about also caching common `Uint` types like `Uint(8)` or `Uint(16)`, or having a lazily populated table of cached `Uint`s of various sizes, but decided it'd be overkill (esp. since we don't have that today).